src/query/function_score_query.rs

```diff
@@ -0,0 +1,289 @@
+use crate::query::{EnableScoring, Explanation, Query, Scorer, Weight};
+use crate::{DocId, DocSet, Score, SegmentReader, TantivyError};
+use std::fmt;
+use std::sync::Arc;
+
+type Function = Arc<dyn Fn(&SegmentReader, Score, DocId) -> Score + Send + Sync + 'static>;
+
+/// A FunctionScoreQuery modifies the score of 
+/// matched documents using a custom function.
+/// ```rust
+/// use tantivy::collector::TopDocs;
+/// use tantivy::query::{FunctionScoreQuery, ConstScoreQuery, AllQuery};
+/// use tantivy::schema::{Schema, TEXT};
+/// use tantivy::{doc, Index, IndexWriter};
+/// use std::sync::Arc;
+///
+/// fn example() -> tantivy::Result<()> {
+///     let mut schema_builder = Schema::builder();
+///     let title = schema_builder.add_text_field("title", TEXT);
+///     let schema = schema_builder.build();
+///     let index = Index::create_in_ram(schema);
+///     {
+///         let mut index_writer: IndexWriter = index.writer(15_000_000)?;
+///         index_writer.add_document(doc!(
+///             title => "The Name of the Wind",
+///         ))?;
+///         index_writer.add_document(doc!(
+///             title => "The Diary of Muadib",
+///         ))?;
+///         index_writer.commit()?;
+///     }
+///     let reader = index.reader()?;
+///     let searcher = reader.searcher();
+///     let MULTIPLIER: f32 = 2.0;
+///     let OFFSET: f32 = 1.0;
+///
+///     {
+///         let query = FunctionScoreQuery::new(
+///             Box::new(ConstScoreQuery::new(Box::new(AllQuery), 0.42)),
+///             Arc::new(move 
+/// |_, score, _| score * MULTIPLIER + OFFSET),
```

Comment:
The formatting seems slightly off here, i.e. the line break is probably unnecessary?